File: backend/core-api/src/main/java/ai/govbiz/core/service/SupportProgramSearchException.java
package ai.govbiz.core.service;

import ai.govbiz.core.client.bizinfo.BizInfoClientException;

public final class SupportProgramSearchException extends RuntimeException {

    public enum Failure {
        NOT_CONFIGURED,
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT
    }

    private final Failure failure;

    private SupportProgramSearchException(Failure failure, String message, Throwable cause) {
        super(message, cause);
        this.failure = failure;
    }

    static SupportProgramSearchException fromClient(BizInfoClientException exception) {
        Failure failure = switch (exception.failure()) {
            case NOT_CONFIGURED -> Failure.NOT_CONFIGURED;
            case UPSTREAM_ERROR -> Failure.UPSTREAM_ERROR;
            case INVALID_RESPONSE -> Failure.INVALID_RESPONSE;
            case UNAVAILABLE -> Failure.UNAVAILABLE;
            case TIMEOUT -> Failure.TIMEOUT;
        };
        return new SupportProgramSearchException(failure, exception.getMessage(), exception);
    }

    public Failure failure() {
        return failure;
    }
}

File: backend/core-api/src/main/kotlin/ai/govbiz/core/service/SupportProgramSearchException.kt
package ai.govbiz.core.service

import ai.govbiz.core.client.bizinfo.BizInfoClientException

class SupportProgramSearchException private constructor(
    val failure: Failure,
    message: String?,
    cause: Throwable?,
) : RuntimeException(message, cause) {
    enum class Failure {
        NOT_CONFIGURED,
        UPSTREAM_ERROR,
        INVALID_RESPONSE,
        UNAVAILABLE,
        TIMEOUT,
    }

    companion object {
        internal fun fromClient(exception: BizInfoClientException): SupportProgramSearchException {
            val failure = when (exception.failure) {
                BizInfoClientException.Failure.NOT_CONFIGURED -> Failure.NOT_CONFIGURED
                BizInfoClientException.Failure.UPSTREAM_ERROR -> Failure.UPSTREAM_ERROR
                BizInfoClientException.Failure.INVALID_RESPONSE -> Failure.INVALID_RESPONSE
                BizInfoClientException.Failure.UNAVAILABLE -> Failure.UNAVAILABLE
                BizInfoClientException.Failure.TIMEOUT -> Failure.TIMEOUT
            }
            return SupportProgramSearchException(failure, exception.message, exception)
        }
    }
}
